File: src/store_canonical_jsonl.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "json"
require "optparse"
require "time"

DEFAULT_OPTIONS = {
  "input" => "data/canonical-items/latest.json",
  "store_dir" => "data/store",
  "date" => nil
}.freeze

def parse_options(argv)
  options = DEFAULT_OPTIONS.dup

  OptionParser.new do |opts|
    opts.banner = "Usage: src/store_canonical_jsonl.rb [options]"
    opts.on("--input PATH", "Canonical items JSON path") { |value| options["input"] = value }
    opts.on("--store-dir PATH", "Store directory") { |value| options["store_dir"] = value }
    opts.on("--date YYYY-MM-DD", "Store partition date") { |value| options["date"] = value }
  end.parse!(argv)

  options
end

def present?(value)
  !(value.nil? || value.to_s.strip.empty?)
end

def partition_date(options)
  value = options["date"]
  return value if present?(value)

  Time.now.strftime("%Y-%m-%d")
end

def jsonl_existing_ids(path)
  return {} unless File.exist?(path)

  ids = {}
  File.foreach(path).with_index(1) do |line, line_number|
    next if line.strip.empty?

    record = JSON.parse(line)
    id = record["id"]
    ids[id] = line_number if present?(id)
  end
  ids
end

def validate_item!(item, index)
  required = %w[id source_id source_type provider title fetched_at content_hash dedupe_key]
  missing = required.reject { |field| present?(item[field]) }
  raise "item #{index} missing required fields: #{missing.join(", ")}" unless missing.empty?
end

def append_jsonl(path, records)
  return if records.empty?

  FileUtils.mkdir_p(File.dirname(path))
  File.open(path, "a") do |file|
    records.each do |record|
      file.write(JSON.generate(record))
      file.write("\n")
    end
  end
end

def store(canonical, options)
  date = partition_date(options)
  store_dir = options["store_dir"]
  items_path = File.join(store_dir, "items", "#{date}.jsonl")
  runs_path = File.join(store_dir, "runs", "#{date}.jsonl")
  stored_at = Time.now.iso8601

  items = canonical.fetch("items")
  existing_ids = jsonl_existing_ids(items_path)
  appended = []
  skipped = []

  items.each_with_index do |item, index|
    validate_item!(item, index)

    if existing_ids.key?(item.fetch("id"))
      skipped << {
        "id" => item.fetch("id"),
        "source_id" => item.fetch("source_id"),
        "reason" => "duplicate_id",
        "existing_line" => existing_ids[item.fetch("id")]
      }
      next
    end

    appended << item.merge(
      "stored_at" => stored_at,
      "store_partition" => date,
      "input_run_id" => canonical["run_id"]
    )
    existing_ids[item.fetch("id")] = existing_ids.length + appended.length
  end

  append_jsonl(items_path, appended)

  run_record = {
    "stored_at" => stored_at,
    "store_partition" => date,
    "input_run_id" => canonical["run_id"],
    "input_adapter" => canonical["input_adapter"],
    "items_path" => items_path,
    "input_items" => items.length,
    "appended_items" => appended.length,
    "skipped_duplicates" => skipped.length,
    "dropped_items" => canonical.dig("summary", "dropped_items"),
    "skipped" => skipped
  }
  append_jsonl(runs_path, [run_record])

  {
    "items_path" => items_path,
    "runs_path" => runs_path,
    "input_items" => items.length,
    "appended_items" => appended.length,
    "skipped_duplicates" => skipped.length
  }
end

def main(argv)
  options = parse_options(argv)
  canonical = JSON.parse(File.read(options["input"]))
  summary = store(canonical, options)

  puts JSON.pretty_generate(summary)
end

main(ARGV)

File: src/agentic-core/agentic_core/pipeline/store_canonical_jsonl.py
import json
from datetime import datetime
from pathlib import Path
from typing import Any


def present(value: Any) -> bool:
    return value is not None and str(value).strip() != ""


def partition_date(options: dict[str, Any]) -> str:
    if present(options.get("date")):
        return str(options["date"])
    return datetime.now().strftime("%Y-%m-%d")


def jsonl_existing_ids(path: Path) -> dict[str, int]:
    if not path.exists():
        return {}
    ids = {}
    for line_number, line in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        if not line.strip():
            continue
        record = json.loads(line)
        if present(record.get("id")):
            ids[record["id"]] = line_number
    return ids


def validate_item(item: dict[str, Any], index: int) -> None:
    required = ["id", "source_id", "source_type", "provider", "title", "fetched_at", "content_hash", "dedupe_key"]
    missing = [field for field in required if not present(item.get(field))]
    if missing:
        raise RuntimeError(f"item {index} missing required fields: {', '.join(missing)}")


def append_jsonl(path: Path, records: list[dict[str, Any]]) -> None:
    if not records:
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as handle:
        for record in records:
            handle.write(json.dumps(record, ensure_ascii=False, separators=(",", ":")))
            handle.write("\n")


def store(canonical: dict[str, Any], options: dict[str, Any], stored_at: str | None = None) -> dict[str, Any]:
    date = partition_date(options)
    store_dir = Path(options["store_dir"])
    items_path = store_dir / "items" / f"{date}.jsonl"
    runs_path = store_dir / "runs" / f"{date}.jsonl"
    stored_at = stored_at or datetime.now().astimezone().isoformat()
    items = canonical["items"]
    existing_ids = jsonl_existing_ids(items_path)
    appended = []
    skipped = []
    for index, item in enumerate(items):
        validate_item(item, index)
        item_id = item["id"]
        if item_id in existing_ids:
            skipped.append({"id": item_id, "source_id": item["source_id"], "reason": "duplicate_id", "existing_line": existing_ids[item_id]})
            continue
        appended.append({**item, "stored_at": stored_at, "store_partition": date, "input_run_id": canonical.get("run_id")})
        existing_ids[item_id] = len(existing_ids) + len(appended)

    append_jsonl(items_path, appended)
    run_record = {
        "stored_at": stored_at,
        "store_partition": date,
        "input_run_id": canonical.get("run_id"),
        "input_adapter": canonical.get("input_adapter"),
        "items_path": str(items_path),
        "input_items": len(items),
        "appended_items": len(appended),
        "skipped_duplicates": len(skipped),
        "dropped_items": canonical.get("summary", {}).get("dropped_items"),
        "skipped": skipped,
    }
    append_jsonl(runs_path, [run_record])
    return {
        "items_path": str(items_path),
        "runs_path": str(runs_path),
        "input_items": len(items),
        "appended_items": len(appended),
        "skipped_duplicates": len(skipped),
    }


def run(input_path: Path, store_dir: Path, date: str | None = None) -> dict[str, Any]:
    canonical = json.loads(input_path.read_text(encoding="utf-8"))
    return store(canonical, {"store_dir": str(store_dir), "date": date})
